### PR TITLE
Resolved Issue#208 - For validation

### PR DESCRIPTION
**Pull request by kameshsampath**
Resolved Issue #208 , since the PVC are created before actual PVs are created, the Volume reference inside the PVC's which are in status `Pending` or 'Lost` are dangling or pointing to __default__ namespace, since default namespace PV's/PVC are bound to each other, the namespace PV/PVC are not bound, hence the required pods are in Creation State for long time.

Approach:

* after new PV are created for PVC in status `Pending` or 'Lost`, we delete the PVC and recreate them to bind it.

Another thought would be to refrain from creating PVC's during deploy ? ( I am not sure this is valid  thought)

Sample output 
https://gist.github.com/kameshsampath/a70ce3ff57d1f552a2434d525a6a2118

### CODE REVIEW

File: cmds/volumes.go [hunk 1]
```diff
@@ -63,9 +94,10 @@ func NewCmdVolumes(f *cmdutil.Factory) *cobra.Command {
 	return cmd
 }
 
-func findPendingPVS(c *k8sclient.Client, ns string) (bool, []string) {
+func findPendingPVS(c *k8sclient.Client, ns string) (bool, *api.PersistentVolumeClaimList, []string) {
```
Comment (jimmidyson) [inline cmds/volumes.go]: Nit: `findPendingPVs`

File: cmds/volumes.go [hunk 2]
```diff
@@ -76,20 +108,25 @@ func findPendingPVS(c *k8sclient.Client, ns string) (bool, []string) {
...
 
 func createPV(c *k8sclient.Client, ns string, pvcNames []string, sshCommand string) (Result, error) {
 
 	for _, pvcName := range pvcNames {
-		hostPath := "/data/" + pvcName
+		paths := []string{"/data", ns, pvcName}
+		strs := []string{ns, pvcName}
+		nsPvcName := strings.Join(strs, "-")
+
+		hostPath := strings.Join(paths, "/")
```
Comment (jimmidyson) [inline cmds/volumes.go]: Use [`filepath.Join`](https://golang.org/pkg/path/filepath/#Join)

File: cmds/volumes.go [hunk 3]
```diff
@@ -76,20 +108,25 @@ func findPendingPVS(c *k8sclient.Client, ns string) (bool, []string) {
...
+	return false, nil, nil
 }
 
 func createPV(c *k8sclient.Client, ns string, pvcNames []string, sshCommand string) (Result, error) {
 
 	for _, pvcName := range pvcNames {
-		hostPath := "/data/" + pvcName
+		paths := []string{"/data", ns, pvcName}
+		strs := []string{ns, pvcName}
+		nsPvcName := strings.Join(strs, "-")
```
Comment (jimmidyson) [inline cmds/volumes.go]: Simply `nsPvcName := ns + "-" + pvcName`

File: cmds/volumes.go [hunk 4]
```diff
@@ -76,20 +108,25 @@ func findPendingPVS(c *k8sclient.Client, ns string) (bool, []string) {
...
-	return false, nil
+	return false, nil, nil
 }
 
 func createPV(c *k8sclient.Client, ns string, pvcNames []string, sshCommand string) (Result, error) {
 
 	for _, pvcName := range pvcNames {
-		hostPath := "/data/" + pvcName
+		paths := []string{"/data", ns, pvcName}
+		strs := []string{ns, pvcName}
```
Comment (jimmidyson) [inline cmds/volumes.go]: Remove - redundant.

File: cmds/volumes.go (hunk 2 above)
Comment (jimmidyson) [inline cmds/volumes.go]: Also move this under `paths := ...` to keep stuff together (better readability).

File: cmds/volumes.go [hunk 5]
```diff
@@ -49,12 +51,41 @@ func NewCmdVolumes(f *cmdutil.Factory) *cobra.Command {
...
+					for _, item := range items {
+						pvcName := item.ObjectMeta.Name
+						status := item.Status.Phase
+						if status == "Pending" || status == "Lost" {
+							err = c.PersistentVolumeClaims(ns).Delete(pvcName)
+							if err != nil {
+								util.Infof("Error deleting PVC %s\n", pvcName)
+							} else {
+								util.Infof("Recreating PVC %s\n", pvcName)
+								strs := []string{ns, pvcName}
```
Comment (jimmidyson) [inline cmds/volumes.go]: As below, this is redundant.

File: cmds/volumes.go [hunk 6]
```diff
@@ -49,12 +51,41 @@ func NewCmdVolumes(f *cmdutil.Factory) *cobra.Command {
...
+								c.PersistentVolumeClaims(ns).Create(&api.PersistentVolumeClaim{
+									ObjectMeta: api.ObjectMeta{
+										Name:      pvcName,
+										Namespace: ns,
+									},
+									Spec: api.PersistentVolumeClaimSpec{
+										VolumeName:  strings.Join(strs, "-"),
```
Comment (jimmidyson) [inline cmds/volumes.go]: `VolumeName: ns + "-" + pvcName`

File: cmds/volumes.go [hunk 7]
```diff
@@ -49,12 +51,41 @@ func NewCmdVolumes(f *cmdutil.Factory) *cobra.Command {
...
+										AccessModes: []api.PersistentVolumeAccessMode{api.ReadWriteOnce},
+										Resources: api.ResourceRequirements{
+											Requests: api.ResourceList{
+												api.ResourceName(api.ResourceStorage): resource.MustParse("5Gi"),
```
Comment (jimmidyson) [inline cmds/volumes.go]: `5Gi` - where does this amount come from? Is it enough? Is it too much? @rawlingsj thoughts?

File: cmds/volumes.go [hunk 8]
```diff
@@ -76,20 +108,25 @@ func findPendingPVS(c *k8sclient.Client, ns string) (bool, []string) {
 		for _, item := range items {
 			status := item.Status.Phase
 			if status == "Pending" || status == "Lost" {
-				pendingClaimNames = append(pendingClaimNames, item.ObjectMeta.Name)
+				pvcName := item.ObjectMeta.Name
```
Comment (jimmidyson) [inline cmds/volumes.go]: Redundant assignment, just use `item.ObjectMeta.Name` below.

File: cmds/volumes.go (hunk 7 above)
Comment (rawlingsj) [inline cmds/volumes.go]: 5 sounds a little high, shall we default to 1Gi instead?

File: cmds/volumes.go (hunk 5 above)
Comment (kameshsampath) [inline cmds/volumes.go]: i guess these comments are on volumes.go

Comment (kameshsampath) [inline cmds/volumes.go]: Fixed

File: cmds/volumes.go (hunk 6 above)
Comment (kameshsampath) [inline cmds/volumes.go]: Fixed

File: cmds/volumes.go (hunk 7 above)
Comment (kameshsampath) [inline cmds/volumes.go]: just took the value as it is not logic behind it. now changed back to 1Gi

File: cmds/volumes.go (hunk 1 above)
Comment (kameshsampath) [inline cmds/volumes.go]: renamed the func

File: cmds/volumes.go (hunk 8 above)
Comment (kameshsampath) [inline cmds/volumes.go]: removed variable `pvcName` and used item.ObjectMeta.Name where needed

File: cmds/volumes.go (hunk 4 above)
Comment (kameshsampath) [inline cmds/volumes.go]: fixed

File: cmds/volumes.go (hunk 3 above)
Comment (kameshsampath) [inline cmds/volumes.go]: fixed

File: cmds/volumes.go (hunk 2 above)
Comment (kameshsampath) [inline cmds/volumes.go]: used filepath.Join

**Comment by jimmidyson** [comment]
LGTM. Let's merge then I will update deps & switch to `Lost` const. Thanks!

[merge]

**Pull request merged**